rust/libnewsboat/src/matcher.rs

```diff
@@ -21,16 +25,960 @@ impl Matcher {
     ///
     /// If `input` can't be parsed, returns a user-readable error.
     pub fn parse(input: &str) -> Result<Matcher, String> {
-        unimplemented!()
+        let expr = filterparser::parse(input).map_err(|e| match e {
+            filterparser::Error::TrailingCharacters(tail) => {
+                format!("Parse error: trailing characters: {}", tail)
+            }
+            filterparser::Error::AtPos(pos) => format!("Parse error at position {}", pos),
+        })?;
+
+        Ok(Matcher {
+            expr,
+            text: input.to_string(),
+        })
     }
 
     /// Check if given matchable `item` matches the filter.
     pub fn matches(&self, item: &impl Matchable) -> Result<bool, MatcherError> {
-        unimplemented!()
+        evaluate_expression(&self.expr, item)
     }
 
     /// The filter expression from which this `Matcher` was constructed.
     pub fn get_expression(&self) -> &str {
         &self.text
     }
 }
+
+impl Operator {
+    fn apply(&self, attr: &str, value: &Value) -> Result<bool, MatcherError> {
+        match self {
+            Operator::Equals => match value {
+                Value::Str(ref s) => Ok(attr == s),
+                Value::Int(i) => match attr.parse::<i32>() {
+                    Ok(attr) => Ok(attr == *i),
+                    _ => Ok(false),
+                },
+                _ => Ok(false),
+            },
+            Operator::NotEquals => Operator::Equals
+                .apply(attr, value)
+                .and_then(|result| Ok(!result)),
+            Operator::RegexMatches => match value {
+                Value::Str(ref pattern) => {
+                    match Regex::new(
+                        pattern,
+                        CompFlags::EXTENDED | CompFlags::IGNORE_CASE | CompFlags::NO_SUB,
+                    ) {
+                        Ok(regex) => {
+                            let mut result = false;
+                            let max_matches = 1;
+                            if let Ok(matches) =
+                                regex.matches(attr, max_matches, MatchFlags::empty())
+                            {
+                                // Ok with non-empty Vec inside means a match was found
+                                result = !matches.is_empty();
+                            }
+                            Ok(result)
+                        }
+
+                        Err(errmsg) => Err(MatcherError::InvalidRegex {
+                            regex: pattern.clone(),
+                            errmsg,
+                        }),
+                    }
+                }
+                _ => Ok(false),
+            },
+            Operator::NotRegexMatches => Operator::RegexMatches
+                .apply(attr, value)
+                .and_then(|result| Ok(!result)),
+            Operator::LessThan => match value {
+                Value::Int(i) => Ok(attr.parse::<i32>().unwrap() < *i),
```

Comment:
It will. As I explain in the commit message for https://github.com/newsboat/newsboat/pull/900/commits/11f7311a31754f904c5913461f9f4039ddc142bf, I only considered the "happy path" for now. This will be fixed once I extended the tests further.

I should've copied commit messages to the first post in the PR, so you don't have to hunt for that info. Will try to remember this the next time.

Reply:
I did read through those commit messages but apparently forgot about the relevant paragraph by the time I arrived at this piece of code.

> This will be fixed once I extended the tests further.

Sounds good.